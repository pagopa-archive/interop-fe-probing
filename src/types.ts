export interface ServiceMainData {
  /**
   * name of the service
   */
  eserviceName: string
  /**
   * name of the producer
   */
  producerName: string
  /**
   * version of the service
   */
  versionNumber: string
  /**
   * polling frequency for the service
   */
  pollingFrequency: number
}

export interface ServiceProbingData {
  /**
   * monitoring state
   */
  probingEnabled: boolean
  /**
   * state of the e-service
   */
  state: string
  /**
   * last detection date
   */
  responseReceived: string | null
  /**
   * active status of the e-service
   */
  eserviceActive: boolean
}

export interface ServiceValuesType {
  /**
   * response time for the service
   */
  responseTime: number
  /**
   * check time for the service
   */
  time: string
}

export interface ServiceFailuresType {
  /**
   * status of the service
   */
  status: string
  /**
   * check time for the service
   */
  time: string
}

export interface ServicePercentagesType {
  /**
   * status of the service
   */
  status: string
  /**
   * percentages of the service
   */
  value: number
}

export interface ServiceStatisticsData {
  percentages: Array<ServicePercentagesType>
  values: Array<ServiceValuesType>
  failures: Array<ServiceFailuresType>
}
