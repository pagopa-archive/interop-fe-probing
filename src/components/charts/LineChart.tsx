import { useEffect, useRef } from "react";
import {
  select,
  scaleLinear,
  scaleTime,
  axisBottom,
  axisLeft,
  line,
  timeFormat,
  curveBasis,
  max,
  curveNatural,
  curveBundle,
  curveBasisOpen,
  curveCardinal,
  ScaleTime,
  ScaleLinear,
} from "d3";
import _ from "lodash";

const curveType = curveBasis;
const DASH_LENGTH = 2;
const DASH_SEPARATOR_LENGTH = 2;

interface IService {
  name: string;
  status: string;
  response_time: number;
  check_time: string;
}

interface IProps {
  data: Array<IService>;
}

// margin convention often used with D3
const margin = { top: 20, right: 30, bottom: 40, left: 60 };
const width = 700 - margin.left - margin.right;
const height = 350 - margin.top - margin.bottom;

const color: { [key: string]: string } = {
  "OK": "#17324D",
  "n/d": "#A2ADB8",
  "KO": "#FE6666",
};

export const LineChart: React.FC<IProps> = ({ data }) => {
  //refs
  const d3svg = useRef(null);

  useEffect(() => {
    if (data && d3svg.current) {
      // append the svg object to our svg and set the width and height
      let svg = select(d3svg.current)
        .attr("width", width + margin.left + margin.right)
        .attr("height", height + margin.top + margin.bottom);

      // draw header
      svg
        .append("g")
        .attr("class", "bar-header")
        .attr("transform", `translate(${margin.left}, ${margin.top})`)
        .append("text")
        .append("tspan")
        .attr("font-family", "Titillium Web")
        .attr("font-size", "18px")
        .attr("color", "#17324D")
        .attr("font-weight", "700")
        .text("Tempi di risposta dellâ€™e-service");

      // create axis
      const xScale = scaleTime()
        .range([0, width])
        .domain([
          // get the previous date of the first check_date in the array
          // because the chart is not including the first one
          new Date(
            new Date(data[0].check_time).setDate(
              new Date(data[0].check_time).getDate() - 1
            )
          ),
          new Date(data[data.length - 1].check_time),
        ]);

      const yScale = scaleLinear()
        .domain([
          0,
          max(data, (d: IService) => {
            return d.response_time || null;
          }),
        ] as Array<number>)

        .range([height, 0]);

      // create common element where to append the axises
      svg
        .append("g")
        .attr("class", "scales")
        .attr("transform", `translate(${margin.left}, ${margin.top * 1.5})`);

      // add x axis
      svg
        .select(".scales")
        .append("g")
        // change ticks style
        .attr("transform", `translate(0, ${height})`)
        .attr("font-family", "Poppins")
        .attr("opacity", 0.4)
        .call(
          axisBottom(xScale)
            .tickPadding(15)
            .tickSize(0)
            // format the ticks labels to be dd / mm
            .tickFormat((date: any) => {
              return timeFormat("%d / %m")(new Date(date));
            })
        )

        .call((g) => {
          // create the grid vertical lines
          g.selectAll(".tick line")
            .attr("transform", `translate(0, ${-height})`)
            .clone()
            .attr("y2", height)
            .attr("stroke-opacity", 0.05);
          // hide the axis line
          g.select(".domain").remove();
        });

      // add y axis
      svg
        .select(".scales")
        .append("g")
        // change ticks style
        .attr("font-family", "Poppins")
        .attr("opacity", 0.4)
        .call(
          axisLeft(yScale)
            .tickPadding(15)
            .tickSize(0)
            // format the ticks labels adding ms in the end
            .tickFormat((data: any) => {
              return `${data}ms`;
            })
        )
        // create the horizontal grid lines
        .call((g) => {
          g.selectAll(".tick line")
            .clone()
            .attr("x2", width)
            .attr("stroke-opacity", 0.05);
          // hide the axis line
          g.select(".domain").remove();
        });

      // create the line
      const valueLine = line<IService>()
        // .curve(curveType)
        .x((d: IService) => {
          return xScale(new Date(d.check_time));
        })
        .y((d: IService) => yScale(d.response_time));

      // color the line in different colors depending on the status
      svg
        .append("linearGradient")
        .attr("id", "linearGradient")
        .attr("gradientUnits", "userSpaceOnUse")
        .attr("x1", 0)
        .attr("x2", width)
        .selectAll("stop")
        .data(data)

        .join("stop")
        .attr("offset", (d: IService) => xScale(new Date(d.check_time)) / width)
        .attr("stop-color", (d: IService) => {
          return color[d.status];
        })
        .attr("stop-opacity", 1);

      // add the line to the chart
      svg
        .append("path")
        .attr("class", "line")
        .datum(data)
        .attr("fill", "none")
        .attr("stroke", "url(#linearGradient)")
        .attr("stroke-width", 1)
        .attr("transform", `translate(${margin.left}, ${margin.top * 1.5})`)
        .attr("d", valueLine);
      // .attr("stroke-dasharray", function (d) {
      //   return getDashArray(d, this, xScale, yScale);
      // });
    }
  }, []);

  // // get the dash array for the stroke-dasharray property
  // const getDashArray = (
  //   data: Array<IService>,
  //   path: SVGPathElement,
  //   xScale: ScaleTime<number, number, never>,
  //   yScale: ScaleLinear<number, number, never>
  // ): Array<number | string> | null => {
  //   const dashedRanges = getDashedRanges(data);
  //   if (dashedRanges.length === 0) return null;
  //   const lengths = data.map((d) =>
  //     getLengthForPoint(
  //       {
  //         x: xScale(new Date(d.check_time)),
  //         y: yScale(new Date(d.response_time)),
  //       },
  //       path
  //     )
  //   );
  //   const dashArr = buildDashArray(dashedRanges, lengths);
  //   return dashArr;
  // };

  // // get array with pointes between which points we have dashed parts
  // // for example betwewen the elements with indexes 3 and 4, and 8 and 9 from tha data array
  // // [{ start: 3, end: 4 }, { start: 8, end: 9 }]
  // const getDashedRanges = (data: Array<IService>): Array<any> => {
  //   const hasOpenRange = (arr: Array<any>) =>
  //     _.last(arr) && !("end" in _.last(arr));
  //   const lastIndex = data.length - 1;
  //   return _.reduce(
  //     data,
  //     (res: Array<any>, d: IService, i: number) => {
  //       const isRangeStart = !hasOpenRange(res) && isDashed(d);
  //       if (isRangeStart) res.push({ start: Math.max(0, i) });

  //       const isRangeEnd =
  //         hasOpenRange(res) && (!isDashed(d) || i === lastIndex);
  //       if (isRangeEnd) res[res.length - 1].end = i - 1;
  //       return res;
  //     },
  //     []
  //   );
  // };

  // // the length from the start of x axis to the point
  // function getLengthForPoint(
  //   point: { x: number; y: number },
  //   thePath: SVGPathElement
  // ) {
  //   let pathLength = thePath.getTotalLength();
  //   let precision = 100;
  //   let division = pathLength / precision;
  //   let theRecord = pathLength;
  //   let theSegment: number = 0;

  //   for (let i = 0; i < precision; i++) {
  //     // get a point on the path for thia distance
  //     let _p = thePath.getPointAtLength(i * division);
  //     // get the distance between the new point _p and the point p
  //     let theDistance = dist(_p, point);
  //     if (theDistance < theRecord) {
  //       // if the distance is smaller than the record set the new record
  //       theRecord = theDistance;
  //       theSegment = i;
  //     }
  //   }
  //   return theSegment * division;
  // }

  // // math formula for calculating the distance between two points
  // function dist(p1: DOMPoint, p2: { x: number; y: number }) {
  //   let dx = p2.x - p1.x;
  //   let dy = p2.y - p1.y;
  //   return Math.sqrt(dx * dx + dy * dy);
  // }

  // // builds the whole array for the stroke-dasharray property
  // function buildDashArray(
  //   dashedRanges: { end: number; start: number }[],
  //   lengths: Array<number>
  // ) {
  //   const dashArr = _.reduce(
  //     dashedRanges,
  //     (res: Array<any>, { start, end }, i) => {
  //       const prevEnd = i === 0 ? 0 : dashedRanges[i - 1].end;
  //       const normalSegment = lengths[start] - lengths[prevEnd];
  //       const dashedSegment = getDashedSegment(lengths[end] - lengths[start]);
  //       return res.concat(normalSegment, dashedSegment);
  //     },
  //     []
  //   );
  //   return dashArr;
  // }

  // // get the part of the stroke-dasharray property for some of the dashed parts
  // function getDashedSegment(length: number): string | undefined {
  //   const totalDashLen = DASH_LENGTH + DASH_SEPARATOR_LENGTH;
  //   const dashCount = Math.floor(length / totalDashLen);
  //   const result: any[] = _.range(dashCount).map(() => {
  //     return DASH_SEPARATOR_LENGTH + "," + DASH_LENGTH;
  //   });
  //   return result.concat(length - dashCount * totalDashLen).join(",");
  // }

  // // check if this point is part of a dashed segment
  // function isDashed(d: IService) {
  //   return d.status != "OK";
  // }

  return (
    <svg
      width={width + margin.left + margin.right}
      height={height + margin.top + margin.bottom}
      ref={d3svg}
    ></svg>
  );
};
