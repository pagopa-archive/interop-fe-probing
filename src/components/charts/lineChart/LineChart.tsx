import {
  scaleLinear,
  scaleTime,
  line,
  timeFormat,
  max,
  Line,
  ScaleLinear,
  ScaleTime,
  curveCatmullRom,
} from 'd3'
import { useTranslation } from 'react-i18next'
import { ServicePerformancesType, ServiceFailuresType } from '../../../types'
import { Grid } from '@mui/material'
import { subMonths, isAfter, addMonths } from 'date-fns'

const curveType = curveCatmullRom.alpha(0.5)

interface IProps {
  data: Array<ServicePerformancesType>
  failures: Array<ServiceFailuresType>
  startDate?: string
  endDate?: string
}

// margin convention often used with D3
const margin = { top: 20, right: 30, bottom: 80, left: 60 }
const width = 700 - margin.left - margin.right
const height = 400 - margin.top - margin.bottom

const color: { [key: string]: string } = {
  OK: '#17324D',
  'N/D': '#A2ADB8',
  KO: '#FE6666',
}

/**
 * Line chart component
 * @component
 */
export const LineChart: React.FC<IProps> = ({ data, failures, startDate, endDate }) => {
  const { t } = useTranslation(['detailsPage'])

  // header of the chart
  const header: JSX.Element = (
    <g className="bar-header" transform={`translate(10, ${margin.top})`}>
      <text>
        <tspan fontFamily="Titillium Web" fontSize="18px" color="#17324D" fontWeight="700">
          {t('lineChartTitle')}
        </tspan>
      </text>
    </g>
  )

  const minTime = startDate
    ? new Date(startDate)
    : endDate
    ? subMonths(new Date(endDate), 3)
    : new Date(new Date().getTime() - 1 * 24 * 60 * 60 * 1000)

  const maxTime = endDate
    ? new Date(endDate)
    : startDate && !isAfter(addMonths(new Date(startDate), 3), new Date())
    ? addMonths(new Date(startDate), 3)
    : new Date()

  //x scale
  const x: ScaleTime<number, number, never> = scaleTime()
    .range([0, width])
    .domain([minTime, maxTime])

  //y scale
  const y: ScaleLinear<number, number, never> = scaleLinear()
    .domain([
      0,
      // if we have null for example when the array is empty the default value is 10
      max(data, (d: ServicePerformancesType) => {
        return d.responseTime || null
      }) || 10,
    ] as Array<number>)

    .range([height, 0])

  // create the x ticks and create from them the vertical lines of the grid
  const xTicks: JSX.Element = (
    <g transform={`translate(0, ${height})`} opacity="40%" fontSize={10} textAnchor="middle">
      {x.ticks().map((d: Date) => (
        <g opacity="1" className="tick" transform={`translate(${x(d)}, 0)`} key={d.getTime()}>
          <line
            y2={0}
            transform={`translate(0, ${-height})`}
            strokeOpacity="0.2"
            stroke="currentColor"
          />
          <line
            y2={height}
            transform={`translate(0, ${-height})`}
            strokeOpacity="0.2"
            stroke="currentColor"
          />
          <text y="15" dy="0.71em">
            {timeFormat('%d / %m')(new Date(d))}
          </text>
          <text y="35">{timeFormat('%H:%M')(new Date(d))}</text>
        </g>
      ))}
    </g>
  )

  // create the y ticks and create from them the horizontal lines of the grid
  const yTicks: JSX.Element = (
    <g opacity="40%" fontSize={10} textAnchor="end">
      {y.ticks().map((d) => (
        <g key={d} opacity="1" className="tick" transform={`translate(0, ${y(d)})`}>
          <line stroke="currentColor" x2={0} strokeOpacity="0.2" />
          <line stroke="currentColor" x2={width} strokeOpacity="0.2" />
          <text x="-15" dy="0.32em">{`${d}ms`}</text>
        </g>
      ))}
    </g>
  )

  //line generator: each point is [x(d.time), y(d.responseTime)] where d is an element in data array
  // and x, y are scales (e.g. x(10) returns pixel value of 10 scaled by x)
  const createLine: Line<ServicePerformancesType> = line<ServicePerformancesType>()
    .curve(curveType)
    .defined(function (d) {
      return d.responseTime !== 0
    })
    .x((d: ServicePerformancesType) => {
      return x(new Date(d.time))
    })
    .y((d: ServicePerformancesType) => y(d.responseTime ? d.responseTime : 0))

  const failuresPoints = (
    <g className="points" transform={`translate(${margin.left}, ${margin.top * 3})`}>
      {failures.map((failure) => (
        <circle
          key={x(new Date(failure.time))}
          cx={x(new Date(failure.time))}
          r="3"
          fill={color[failure.status]}
        ></circle>
      ))}
    </g>
  )

  // create the x ticks for the failures chart
  const xTicksFailures: JSX.Element = (
    <g opacity="40%" fontSize={10} textAnchor="middle">
      {x.ticks().map((d: Date) => (
        <g opacity="1" className="tick" transform={`translate(${x(d)}, 0)`} key={d.getTime()}>
          <text y="15" dy="0.71em">
            {timeFormat('%d / %m')(new Date(d))}
          </text>
          <text y="35">{timeFormat('%H:%M')(new Date(d))}</text>
        </g>
      ))}
    </g>
  )

  // header of the chart
  const failuresHeader: JSX.Element = (
    <g className="failures-header" transform={`translate(10, ${margin.top})`}>
      <text>
        <tspan fontFamily="Titillium Web" fontSize="18px" color="#17324D" fontWeight="700">
          {t('lineChartFailuresTitle')}
        </tspan>
      </text>
    </g>
  )

  return (
    <Grid container direction={'column'} rowGap={4}>
      <Grid item>
        <svg
          className="line-chart-container"
          width={width + margin.left + margin.right}
          height={height + margin.top + margin.bottom}
          role="img"
        >
          {header}
          <g className="scales" transform={`translate(${margin.left}, ${margin.top * 3})`}>
            {xTicks}
            {yTicks}
          </g>
          <path
            className="line"
            stroke="#17324D"
            strokeWidth="1"
            fill="none"
            d={createLine(data) || undefined}
            transform={`translate(${margin.left}, ${margin.top * 3})`}
          />
        </svg>
      </Grid>

      <Grid item>
        <svg
          className="line-chart-container-failures"
          width={width + margin.left + margin.right}
          height={100}
          role="img"
        >
          {failuresHeader}
          <g className="scales" transform={`translate(${margin.left}, ${margin.top * 3})`}>
            {xTicksFailures}
            <g fontFamily="Poppins" opacity={0.4} fontSize={10} textAnchor="end">
              <g opacity="1" className="tick">
                <line stroke="currentColor" x2={0} strokeOpacity="0.2" />
                <line stroke="currentColor" x2={width} strokeOpacity="0.2" />
              </g>
            </g>
          </g>
          {failuresPoints}
        </svg>
      </Grid>
    </Grid>
  )
}
