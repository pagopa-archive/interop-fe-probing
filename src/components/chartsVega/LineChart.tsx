import { VegaLite, VisualizationSpec } from "react-vega";

const mainSpec: VisualizationSpec = {
  "title": {
    // the title of the chart
    "text": "Tempi di risposta dellâ€™e-service",
    // align the title in the start; by default it is in the middle
    "anchor": "start",
  },
  "encoding": {
    "x": {
      // connected field from the values properties
      "field": "check_time",
      // "quantitative" if the field is a number
      // "nominal" if the field is a string
      // "temporal" if the field is a date time object
      "type": "temporal",
      // if the title is null it is removed
      "title": null,
      "axis": {
        "ticks": false,
        // so the labels have opacity
        "labelOpacity": 0.6,
        // so the labels in the x axe is not rotated
        "labelAngle": 0,
        // so the labels to be formatted as dd / MM
        "labelExpr": "date(datum.value) + ' / ' + (month(datum.value)+1)",
      },
    },
    "y": {
      "field": "response_time",
      "type": "quantitative",
      "title": null,
      "axis": {
        "grid": true,
        "labelOpacity": 0.6,
        "labelExpr": "datum.label + 'ms'",
      },
    },
  },
  // width and height of the chart
  "width": 400,
  "height": 200,
  // change the color of the x and y axis
  "config": { "axis": { "domainColor": "#ddd", "tickColor": "#ddd" } },
};

interface IService {
  name: string;
  status: string;
  response_time: number;
  check_time: string;
}

interface IProp {
  data: Array<IService>;
}

const LineChart = ({ data }: IProp) => {
  // crete specification layer from the given array of data
  const createLayer = (layerData: any) => {
    const status = layerData[0].status;
    return {
      "data": {
        "values": [...layerData],
        // type of the data, it can be "json", "csv", "tsv", "dsv"
      },
      "mark": {
        "type": "line",
        // so the line is dashed if the the status is KO or n/d
        "strokeDash": status === "KO" || status === "n/d" ? [4, 1] : null,
        // change the colors for the differnt status cases
        ...(status === "KO" && { color: "#FE6666" }),
        ...(status === "OK" && { color: "#17324D" }),
        ...(status === "n/d" && { color: "#A2ADB8" }),
        // so the lines don't have sharp edges
        "interpolate": "basis",
      },
    };
  };

  // split the data to different arrays(layers) by status and dates so we have diffent lines(layers)
  const createLayers = (items: any[]) => {
    const layers: any[] = [];
    let currentData: any[] = [];
    items.forEach((item, index) => {
      if (index === 0) {
        currentData.push(item);
      } else if (item.status === data[index - 1].status) {
        currentData.push(item);
        if (index + 1 == items.length) {
          layers.push(createLayer(currentData));
        }
      } else {
        layers.push(createLayer(currentData));
        currentData = [item];
      }
    });
    return layers;
  };

  // create specification with the main spec and the created layers
  // and return the ready chart
  const createChart = () => {
    const layers = createLayers(data);

    let spec: VisualizationSpec = {
      ...mainSpec,
      "layer": [...layers],
    };

    return <VegaLite spec={spec} actions={false} />;
  };

  return createChart();
};
export default LineChart;
