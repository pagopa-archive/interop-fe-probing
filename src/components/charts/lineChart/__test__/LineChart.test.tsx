import { describe, expect, test } from 'vitest'
import { render } from '@testing-library/react'
import { LineChart } from '../LineChart'
import { ServiceValuesType } from '../../../../types'

const data: Array<ServiceValuesType> = [
  {
    status: 'OK',
    responseTime: 50,
    time: '2023-12-01T14:20:15.995Z',
  },
  {
    status: 'OK',
    responseTime: 70,
    time: '2023-12-03T14:20:15.995Z',
  },
  {
    status: 'OK',
    responseTime: 0,
    time: '2023-12-04T14:20:15.995Z',
  },
]

describe('LineChart', () => {
  test('render component', () => {
    const { container } = render(<LineChart data={data} />)
    expect(container).toMatchSnapshot()
  })
})
