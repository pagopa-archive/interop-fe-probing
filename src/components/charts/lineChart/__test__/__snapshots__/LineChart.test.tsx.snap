// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LineChart > render component 1`] = `
<div>
  <svg
    class="line-chart-container"
    height="400"
    role="img"
    width="700"
  >
    <g
      class="bar-header"
      transform="translate(10, 20)"
    >
      <text>
        <tspan
          color="#17324D"
          font-family="Titillium Web"
          font-size="18px"
          font-weight="700"
        >
          lineChartTitle
        </tspan>
      </text>
    </g>
    <g
      class="scales"
      transform="translate(60, 60)"
    >
      <g
        font-family="Poppins"
        font-size="10"
        opacity="0.4"
        text-anchor="middle"
        transform="translate(0, 300)"
      >
        <g
          class="tick"
          opacity="1"
          transform="translate(22.554950038580248, 0)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="300"
          />
          <text
            dy="0.71em"
            y="15"
          >
            01 / 12
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(73.38828337191359, 0)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="300"
          />
          <text
            dy="0.71em"
            y="15"
          >
            02 / 12
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(124.2216167052469, 0)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="300"
          />
          <text
            dy="0.71em"
            y="15"
          >
            02 / 12
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(175.05495003858027, 0)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="300"
          />
          <text
            dy="0.71em"
            y="15"
          >
            02 / 12
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(225.8882833719136, 0)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="300"
          />
          <text
            dy="0.71em"
            y="15"
          >
            02 / 12
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(276.7216167052469, 0)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="300"
          />
          <text
            dy="0.71em"
            y="15"
          >
            03 / 12
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(327.5549500385803, 0)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="300"
          />
          <text
            dy="0.71em"
            y="15"
          >
            03 / 12
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(378.38828337191353, 0)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="300"
          />
          <text
            dy="0.71em"
            y="15"
          >
            03 / 12
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(429.2216167052469, 0)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="300"
          />
          <text
            dy="0.71em"
            y="15"
          >
            03 / 12
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(480.0549500385803, 0)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="300"
          />
          <text
            dy="0.71em"
            y="15"
          >
            04 / 12
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(530.8882833719135, 0)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="300"
          />
          <text
            dy="0.71em"
            y="15"
          >
            04 / 12
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(581.7216167052469, 0)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            transform="translate(0, -300)"
            y2="300"
          />
          <text
            dy="0.71em"
            y="15"
          >
            04 / 12
          </text>
        </g>
      </g>
      <g
        font-family="Poppins"
        font-size="10"
        opacity="0.4"
        text-anchor="end"
      >
        <g
          class="tick"
          opacity="1"
          transform="translate(0, 300)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="610"
          />
          <text
            dy="0.32em"
            x="-15"
          >
            0ms
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0, 278.57142857142856)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="610"
          />
          <text
            dy="0.32em"
            x="-15"
          >
            5ms
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0, 257.14285714285717)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="610"
          />
          <text
            dy="0.32em"
            x="-15"
          >
            10ms
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0, 235.71428571428572)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="610"
          />
          <text
            dy="0.32em"
            x="-15"
          >
            15ms
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0, 214.28571428571428)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="610"
          />
          <text
            dy="0.32em"
            x="-15"
          >
            20ms
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0, 192.85714285714283)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="610"
          />
          <text
            dy="0.32em"
            x="-15"
          >
            25ms
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0, 171.42857142857142)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="610"
          />
          <text
            dy="0.32em"
            x="-15"
          >
            30ms
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0, 150)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="610"
          />
          <text
            dy="0.32em"
            x="-15"
          >
            35ms
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0, 128.57142857142858)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="610"
          />
          <text
            dy="0.32em"
            x="-15"
          >
            40ms
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0, 107.14285714285712)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="610"
          />
          <text
            dy="0.32em"
            x="-15"
          >
            45ms
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0, 85.71428571428571)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="610"
          />
          <text
            dy="0.32em"
            x="-15"
          >
            50ms
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0, 64.28571428571429)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="610"
          />
          <text
            dy="0.32em"
            x="-15"
          >
            55ms
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0, 42.85714285714287)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="610"
          />
          <text
            dy="0.32em"
            x="-15"
          >
            60ms
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0, 21.42857142857142)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="610"
          />
          <text
            dy="0.32em"
            x="-15"
          >
            65ms
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0, 0)"
        >
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="0"
          />
          <line
            stroke="currentColor"
            stroke-opacity="0.05"
            x2="610"
          />
          <text
            dy="0.32em"
            x="-15"
          >
            70ms
          </text>
        </g>
      </g>
    </g>
    <lineargradient
      gradientUnits="userSpaceOnUse"
      id="linearGradient"
      x1="0"
      x2="610"
    >
      <stop
        offset="0"
        stop-color="#17324D"
        stop-opacity="1"
      />
      <stop
        offset="0.6666666666666666"
        stop-color="#17324D"
        stop-opacity="1"
      />
      <stop
        offset="1"
        stop-color="#17324D"
        stop-opacity="1"
      />
    </lineargradient>
    <path
      class="line"
      d="M0,85.714L406.667,0L610,300"
      fill="none"
      stroke="url(#linearGradient)"
      stroke-width="1"
      transform="translate(60, 60)"
    />
  </svg>
</div>
`;
